__tests__/globals_test.ml

```diff
@@ -229,7 +229,7 @@ let () =
     );
 
     describe "timeout should not fail suite" (fun () ->
-      afterAllAsync ~timeout:1 (fun _ ->());
+      afterAllAsync ~timeout:1 (fun (finish) -> Jest.runAllTimers(); finish ());
 
       test "" (fun () -> pass); (* runner will crash if there's no tests *)
     );
```

Comment:
A better approach would be to call expect(finish).toHaveBeenCalled() after Jest.runAllTimers(), but bs-jest does not include bindings for this.  In jest.ml, you added toHaveBeenCalledWith but then commented it out.  I'll take a crack at adding a binding.

Reply:
I still don't see why that would be necessary. It shouldn't matter what happens after a timeout occur. I also highly doubt that these hooks use `setTimeout` directly and is affected by `useFakeTimers`, `runAllTimers` etc., as that would make any non-trivial async test suite a mess.

Reply:
A more actionable comment: This test ought to pass without calling `finish`, or we should make sure the behaviour has actually changed and rewrite the test accordingly.

Reply:
It's hard to say exactly when the current behavior was introduced.  Jest switched from Jasmine to Circus with the 27.0.0 release ([https://github.com/facebook/jest/pull/10686](https://github.com/facebook/jest/pull/10686)).  Jasmine incorporates code that replaces a missing callback fn with an original fn that returns void.  I believe this is the difference.  At any rate, I have added code to the after[All|Each]Async and after[All|Each]Promise code to ensure the timers expire and test that the callback actually executed.  I'll add that change and changes to useFakeTimers() to the PR.  Have a look and let me know whether you think this needs further investigation.

Reply:
See [https://github.com/facebook/jest/issues/6401](https://github.com/facebook/jest/issues/6401).  Jest transitioned from jest-jasmine to jest-circus with release 27.0.0.  The semantics of all async and promise hooks is now consistent - timeouts about the encapsulating suite.

Reply:
I don't see this issue conclusively saying that the behaviour has changed, only the reporting. I.e. that it prints a warning, but doesn't necessarily fail the test. But it might also mean the latter.

So if it is a change in behaviour, the call to `finish` should be removed, we should observe that the test suite does indeed fail and that the stated reason is a timeout, and then add`Skip.` to it so that the whole test run passes. Unfortunately this is the only way to test the test runner itself sometimes.

Reply:
We agree on this.  Please take a look at globals_test.res and runner_test.res in [commit 9a2e52e](https://github.com/glennsl/bs-jest/pull/96/commits/9a2e52ecd9b2e1cfc6e92541dca6273f38f63e11). Timeouts in hooks --- [before | after][Each | All]|[Async | Promise] --- are expected to fail the suite (all hooks) or tests (each hooks).   I ran the tests to confirm this behavior and marked all of them to be skipped once I had confirmed.

Reply:
Ah, sorry I missed that. Looks good then!